File: plop-templates/adapter/payout/src/converter/exit/ExitToProcessResultConverterImpl.java
package com.rbkmoney.{{adapterPayoutPackageCase bank_name}}.converter.exit;

import com.rbkmoney.adapter.bank.payout.spring.boot.starter.converter.ExitStateToProcessResultConverter;
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.AdapterState;
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.EntryStateModel;
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.ExitStateModel;
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.Step;
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.service.IntentService;
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.state.serializer.AdapterStateSerializer;
import com.rbkmoney.adapter.common.model.PollingInfo;
import com.rbkmoney.{{adapterPayoutPackageCase bank_name}}.model.ExitStateModelImpl;
import com.rbkmoney.damsel.msgpack.Value;
import com.rbkmoney.damsel.withdrawals.provider_adapter.Intent;
import com.rbkmoney.damsel.withdrawals.provider_adapter.ProcessResult;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.time.Instant;

@Slf4j
@Component
@RequiredArgsConstructor
public class ExitToProcessResultConverterImpl implements ExitStateToProcessResultConverter<ExitStateModelImpl> {

    private final AdapterStateSerializer stateSerializer;

    private final IntentService intentService;

    @Override
    public ProcessResult convert(ExitStateModelImpl exitStateModel) {
        log.info("ExitState converter. Exit state model: {}", exitStateModel);
        final EntryStateModel entryStateModel = exitStateModel.getEntryStateModel();
        final Step step = entryStateModel.getState().getStep();
        final Intent intent = null;
        if (exitStateModel.getNextState().getPollingInfo() == null) {
            initPollingInfo(exitStateModel);
        }

        AdapterState nextState = exitStateModel.getNextState();
        /**
         * Тут switch() с разными статусами Step
         */
//       пример №1
//        if (step == Step.PAYOUT || step == Step.CHECK) {
//            switch (exitStateModel.getState()) {
//                case SUCCESS:
//                    intent = intentService.getSuccess(exitStateModel);
//                    break;
//                case PENDING:
//                    intent = intentService.getSleep(exitStateModel);
//                    break;
//                case ERROR:
//                case TIMEOUT:
//                    intent = intentService.getFailureByCode(exitStateModel);
//                    break;
//                default:
//                    throw new IllegalStateException("Payout step. Unknown payout state: " + exitStateModel.getState());
//            }
//        } else {
//            throw new IllegalStateException("Unknown step: " + step);
//        }

//       пример №2 (C заполнением TrxInfo)
//        TransactionInfo trxInfo = new TransactionInfo();
//        trxInfo.setTrxId(exitStateModel.getMerchantOrderId());
//        Map<String, String> extra = new HashMap<>();
//        if (exitStateModel.getPaynetOrderId() != null) {
//            extra.put(TrxExtraConstant.PAYNET_ORDER_ID, exitStateModel.getPaynetOrderId());
//        }
//        if (exitStateModel.getSerialNumber() != null) {
//            extra.put(TrxExtraConstant.SERIAL_NUMBER, exitStateModel.getSerialNumber());
//        }
//        trxInfo.setTrxExtra(extra);
//        nextState.setTrxInfo(trxInfo);
//        switch (step) {
//            case PAYOUT:
//                intent = intentService.getSleep(exitStateModel);
//                break;
//            case CHECK:
//                if (exitStateModel.getStatus() != null && APPROVED.equals(exitStateModel.getStatus())) {
//                    intent = Intent.finish(new FinishIntent(FinishStatus.success(new Success())));
//                } else {
//                    intent = createSleepIntent(exitStateModel);
//                }
//                break;
//            default:
//                throw new UnsupportedOperationException("Unknown step : " + step);
//        }

        final ProcessResult processResult = new ProcessResult();
        processResult.setNextState(Value.bin(stateSerializer.writeByte(nextState)));
        processResult.setIntent(intent);

        log.info("ExitState converter. ProcessResult: {}", processResult);

        return processResult;
    }

    private void initPollingInfo(ExitStateModel exitStateModel) {
        PollingInfo pollingInfo = new PollingInfo();
        pollingInfo.setStartDateTimePolling(Instant.now());
        pollingInfo.setMaxDateTimePolling(intentService.extractMaxDateTimeInstant(exitStateModel.getEntryStateModel()));
        exitStateModel.getNextState().setPollingInfo(pollingInfo);
    }

}

File: plop-templates/adapter/payout/src/converter/exit/ExitToProcessResultConverterImpl.kt
package com.rbkmoney.{{adapterPayoutPackageCase bank_name}}.converter.exit

import com.rbkmoney.adapter.bank.payout.spring.boot.starter.converter.ExitStateToProcessResultConverter
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.AdapterState
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.EntryStateModel
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.model.ExitStateModel
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.service.IntentService
import com.rbkmoney.adapter.bank.payout.spring.boot.starter.state.serializer.AdapterStateSerializer
import com.rbkmoney.adapter.common.model.PollingInfo
import com.rbkmoney.{{adapterPayoutPackageCase bank_name}}.model.ExitStateModelImpl
import com.rbkmoney.damsel.msgpack.Value
import com.rbkmoney.damsel.withdrawals.provider_adapter.Intent
import com.rbkmoney.damsel.withdrawals.provider_adapter.ProcessResult
import mu.KotlinLogging
import org.springframework.stereotype.Component
import java.time.Instant

@Component
class ExitToProcessResultConverterImpl(
    private val stateSerializer: AdapterStateSerializer,
    private val intentService: IntentService
) : ExitStateToProcessResultConverter<ExitStateModelImpl> {

    private val log = KotlinLogging.logger { }

    override fun convert(exitStateModel: ExitStateModelImpl): ProcessResult {
        log.info { "ExitState converter. Exit state model: $exitStateModel" }
        val entryStateModel: EntryStateModel = exitStateModel.entryStateModel
        val step = entryStateModel.state.step
        var intent = Intent()
        if (exitStateModel.nextState.pollingInfo == null) {
            initPollingInfo(exitStateModel)
        }
        val nextState: AdapterState = exitStateModel.nextState

// Тут switch() с разными статусами Step
//       пример №1
//        intent = if (step == Step.PAYOUT || step == Step.CHECK) {
//                when (exitStateModel.state) {
//                    SUCCESS -> intentService.getSuccess(exitStateModel)
//                    PENDING -> intentService.getSleep(exitStateModel)
//                    ERROR, TIMEOUT -> intentService.getFailureByCode(exitStateModel)
//                    else -> throw IllegalStateException("Payout step. Unknown payout state: ${entryStateModel.state}")
//                }
//            } else {
//                throw IllegalStateException("Unknown step: $step")
//            }

//       пример №2 (C заполнением TrxInfo)
//        val trxInfo = TransactionInfo()
//        trxInfo.setTrxId(exitStateModel.getMerchantOrderId())
//        val extra: Map<String, String> = HashMap()
//        if (exitStateModel.getPaynetOrderId() != null) {
//            extra.put(TrxExtraConstant.PAYNET_ORDER_ID, exitStateModel.getPaynetOrderId())
//        }
//        if (exitStateModel.getSerialNumber() != null) {
//            extra.put(TrxExtraConstant.SERIAL_NUMBER, exitStateModel.getSerialNumber())
//        }
//        trxInfo.setTrxExtra(extra)
//        nextState.trxInfo = trxInfo
//        when (step) {
//            Step.PAYOUT -> intent = intentService.getSleep(exitStateModel)
//            Step.CHECK -> if (exitStateModel.getStatus() != null && APPROVED.equals(exitStateModel.getStatus())) {
//                intent = Intent.finish(FinishIntent(FinishStatus.success(Success())))
//            } else {
//                intent = createSleepIntent(exitStateModel)
//            }
//            else -> throw UnsupportedOperationException("Unknown step : $step")
//        }
        val processResult = ProcessResult().apply {
            setNextState(Value.bin(stateSerializer.writeByte(nextState)))
            setIntent(intent)
        }
        log.info { "ExitState converter. ProcessResult: $processResult" }
        return processResult
    }

    private fun initPollingInfo(exitStateModel: ExitStateModel) {
        val pollingInfo = PollingInfo().apply {
            startDateTimePolling = Instant.now()
            maxDateTimePolling = intentService.extractMaxDateTimeInstant(exitStateModel.entryStateModel)
        }
        exitStateModel.nextState.pollingInfo = pollingInfo
    }
}
